Tests/NIOCoreTests/SingleStepByteToMessageDecoderTest.swift

```diff
@@ -549,4 +549,35 @@ public final class NIOSingleStepByteToMessageDecoderTest: XCTestCase {
         XCTAssertEqual(2, messageReceiver.count)
         XCTAssertEqual(processor.unprocessedBytes, 0)
     }
+
+    /// Tests re-entrancy by having a nested decoding operation empty the buffer and exit part way
+    /// through the outer processing step
+    func testErrorDuringNestedDecoding() {
+        class ThrowingOnLastDecoder: NIOSingleStepByteToMessageDecoder {
+            /// `ByteBuffer` is the expected type passed in.
+            public typealias InboundIn = ByteBuffer
+            /// `ByteBuffer`s will be passed to the next stage.
+            public typealias InboundOut = ByteBuffer
+
+            public init() { }
+
+            struct DecodeLastError: Error {}
+
+            func decodeLast(buffer: inout NIOCore.ByteBuffer, seenEOF: Bool) throws -> NIOCore.ByteBuffer? {
+                throw DecodeLastError()
+            }
+
+            func decode(buffer: inout NIOCore.ByteBuffer) throws -> NIOCore.ByteBuffer? {
+                ByteBuffer()
+            }
+        }
+
+        let decoder = ThrowingOnLastDecoder()
+        let b2mp = NIOSingleStepByteToMessageProcessor(decoder)
+        XCTAssertNoThrow(try b2mp.process(buffer: ByteBuffer(string: "1\n\n2\n3\n")) { line in
+            XCTAssertThrowsError(try b2mp.finishProcessing(seenEOF: true) { _ in }) { error in
```

Comment:
This test is testing your fix, but unless I misunderstood your comment before the test, I think it's not happening in the way you expected.

`XCTAssertThrowsError` isn't actually throwing, so the closure passed to `process(buffer:messageReceiver:)` won't actually throw when executed. 

This isn't all bad though, because that means that the buffer will be nil after the given closure, which will test your fix as the method will continue without erroring when force-unwrapping the buffer. However, that's not what you were set to test here, since we're not exiting part way.

I think it would make sense to keep this test (if we don't already have one that tests the `messageReceiver` being throwing), but removing the `XCTAssertThrowsError` so that `finishProcessing` actually throws, and we can assert that `process` throws when the inner closure fails. 

However, I would definitely have a separate test to test the actual fix here: to do that you can get rid of the error being thrown on `decodeLast`: we just need to call `finishProcessing` but make sure it doesn't throw to trigger the re-entrancy bug.

Reply:
This was discussed out-of-band, but for future historians...

I did intend the behavior of `XCTAssertThrowsError` to check and then swallow the error thrown from the interior call to processing, but it should've been more obvious. Doing that allowed the interior call to leave `self._buffer` as `nil`.

Whilst investigating this in detail we discovered that `self._buffer` is `nil` because `self._withNonCoWBuffer` would set it to `nil` before calling the decode operation which then threw and then never replaced it. That seems dangerous and we should unconditionally replace the buffer after the decode operation whether or not it throws so I have made that change.

The test is left in place as a regression test with the catching of the error made explicit.